eng/code-quality-reports/src/main/java/com/azure/tools/checkstyle/checks/ServiceClientBuilderCheck.java

```diff
@@ -0,0 +1,155 @@
+
+package com.azure.tools.checkstyle.checks;
+
+import com.puppycrawl.tools.checkstyle.api.AbstractCheck;
+import com.puppycrawl.tools.checkstyle.api.DetailAST;
+import com.puppycrawl.tools.checkstyle.api.TokenTypes;
+
+/**
+ * The @ServiceClientBuilder class should have following rules:
+ *     1) All service client builder should be named <ServiceName>ClientBuilder and annotated with @ServiceClientBuilder.
+ *     2) A property named 'serviceClients'.
+*      3) Has a method 'buildClient()' to build a synchronous client,
+ *     4) Has a method 'buildAsyncClient()' to build a asynchronous client
+ */
+public class ServiceClientBuilderCheck extends AbstractCheck {
+    private static final String SERVICE_CLIENT_BUILDER = "ServiceClientBuilder";
+    private static final String BUILD_CLIENT = "buildClient";
+    private static final String BUILD_ASYNC_CLIENT = "buildAsyncClient";
+
+    private static boolean hasServiceClientBuilderAnnotation;
+    private static boolean hasAsyncClientBuilder;
+    private static boolean hasClientBuilder;
+
+    @Override
+    public void beginTree(DetailAST root) {
+        hasServiceClientBuilderAnnotation = false;
+        hasAsyncClientBuilder = false;
+        hasClientBuilder = false;
+    }
+
+    @Override
+    public void finishTree(DetailAST root) {
+        if (!hasServiceClientBuilderAnnotation) {
+            return;
+        }
+
+        if (!hasAsyncClientBuilder) {
+            log(root, String.format("Every Missing an asynchronous method, ''%s''", BUILD_ASYNC_CLIENT));
+        }
+        if (!hasClientBuilder) {
+            log(root, String.format("Missing a synchronous method, ''%s''", BUILD_CLIENT));
+        }
+    }
+
+    @Override
+    public int[] getDefaultTokens() {
+        return getRequiredTokens();
+    }
+
+    @Override
+    public int[] getAcceptableTokens() {
+        return getRequiredTokens();
+    }
+
+    @Override
+    public int[] getRequiredTokens() {
+        return new int[] {
+            TokenTypes.CLASS_DEF,
+            TokenTypes.METHOD_DEF
+        };
+    }
+
+    @Override
+    public void visitToken(DetailAST token) {
+        switch (token.getType()) {
+            case TokenTypes.CLASS_DEF:
+                final DetailAST serviceClientAnnotationToken = getServiceClientAnnotation(token);
+                final String className = token.findFirstToken(TokenTypes.IDENT).getText();
+
+                hasServiceClientBuilderAnnotation = serviceClientAnnotationToken != null;
+
+                if (hasServiceClientBuilderAnnotation) {
+                    // Checks if the ANNOTATION has property named 'serviceClients'
+                    if (!hasServiceClientsAnnotationProperty(serviceClientAnnotationToken)) {
+                        log(serviceClientAnnotationToken, String.format(
+                            "Annotation @%s should have ''serviceClients'' as property of annotation and should list all of the service clients it can build.",
```

Comment:
You don't need to use %s for ServiceClientBuilder - it's a fixed string.

Reply:
My bad. This log will be removed because of 'serviceClients' is required when they add @ServiceClientBuilder